lib/src/rules/require_trailing_comma.dart

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Use trailing commas for all function calls and definitions.';
+
+const _details = r'''
+
+**DO** use trailing commas for all function calls and definitions unless the
+function call or definition, from the start of the function name up to the
+closing parenthesis, fits in a single line.
+
+**GOOD:**
+```dart
+void run() {
+  method(
+    'does not fit on one line',
+    'test test test test test test test test test test test',
+  );
+}
+```
+
+**BAD:**
+```dart
+void run() {
+  method('does not fit on one line',
+      'test test test test test test test test test test test');
+}
+```
+
+**Exception:** If the final parameter is positional (vs named) and either a
+function literal implemented using curly braces, a literal map, a literal set or
+a literal array. This exception only applies if the final parameter does not fit
+entirely on one line.
+
+''';
+
+class RequireTrailingComma extends LintRule implements NodeLintRule {
+  RequireTrailingComma()
+      : super(
+          name: 'require_trailing_comma',
+          description: _desc,
+          details: _details,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(
+    NodeLintRegistry registry,
+    LinterContext context,
+  ) {
+    var visitor = _Visitor(this);
+    registry
+      ..addCompilationUnit(this, visitor)
+      ..addArgumentList(this, visitor)
+      ..addFormalParameterList(this, visitor)
+      ..addAssertStatement(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  static const _trailingCommaCode = LintCode(
+    'require_trailing_comma',
+    'Missing a required trailing comma.',
+  );
+
+  final LintRule rule;
+
+  LineInfo? _lineInfo;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) => _lineInfo = node.lineInfo;
+
+  @override
+  void visitArgumentList(ArgumentList node) {
+    super.visitArgumentList(node);
+    if (node.arguments.isEmpty) return;
+    _checkTrailingComma(
+      node.leftParenthesis,
+      node.rightParenthesis,
+      node.arguments.last,
+    );
+  }
+
+  @override
+  void visitFormalParameterList(FormalParameterList node) {
+    super.visitFormalParameterList(node);
+    if (node.parameters.isEmpty) return;
+    _checkTrailingComma(
+      node.leftParenthesis,
+      node.rightParenthesis,
+      node.parameters.last,
+    );
+  }
+
+  @override
+  void visitAssertStatement(AssertStatement node) {
```

Comment:
Are we sure we want to treat assert statements as invocations? If so, should we also treat asserts in initializer lists as invocations? If so, you'll also need to visit `AssertInitializer`s.

Reply:
Yes, it should help readability there too. Thanks for pointing out the missing `AssertInitializer` visitor. I've added that and accompanying tests.